### Modeling with to diferent algorithim
- RandomForestClassifier
- xgboost 

In [4]:
# importing lib
import warnings
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
#from logesticRegressionModel import LogesticRegressionModel
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import MinMaxScaler
from sklearn.ensemble import RandomForestClassifier
from sklearn.preprocessing import LabelEncoder


In [5]:
sns.set()
%matplotlib inline
warnings.filterwarnings("ignore")
pd.set_option('display.max_columns', None)
pd.set_option('display.max_colwidth', None)
pd.set_option("expand_frame_repr", False)
pd.set_option('display.float_format', '{:.2f}'.format)

In [6]:
import os
import sys
sys.path.append(os.path.abspath(os.path.join('../scripts')))

In [7]:
df_tripOrder = pd.read_csv('../data/forCusalGraph.csv')

In [8]:
# copy
df_data =df_tripOrder.copy()

In [9]:
# identfying categorical values
non_numeric_columns = list(df_data.select_dtypes(exclude=[np.number]).columns)
print(non_numeric_columns)

['Trip Origin', 'Trip Destination', 'Trip Start Time', 'Trip End Time', 'driver_action']


In [12]:
# labelEncoder 
le = LabelEncoder()
for col in non_numeric_columns:
    df_data[col] = le.fit_transform(df_data[col])
df_data.head(2)

Trip ID  Trip Origin  Trip Destination  Trip Start Time  Trip End Time  Distance  Speed  driver_action  is_weekend  is_holiday
0   392001        11676              3785              275            123     20.98 217.00              0           0           0
1   392001        11676              3785              275            123     20.98 217.00              1           0           0

In [13]:
df_data.describe()

Trip ID  Trip Origin  Trip Destination  Trip Start Time  Trip End Time   Distance      Speed  driver_action  is_weekend  is_holiday
count 1555718.00   1555718.00        1555718.00       1555718.00     1555718.00 1555718.00 1555718.00     1555718.00  1555718.00  1555718.00
mean   419181.69      7466.57           9194.62         12284.80       12277.31       8.47    3705.21           0.98        0.13        0.00
std     14444.35      3684.40           4903.68          6347.94        6352.34       6.34    7806.28           0.13        0.33        0.00
min    392001.00         0.00              0.00             0.00           0.00       0.02      16.00           0.00        0.00        0.00
25%    406752.00      5006.00           5227.00          6853.00        6858.00       4.00    1722.00           1.00        0.00        0.00
50%    421687.00      7699.00           9387.00         13234.00       13236.00       6.90    2608.00           1.00        0.00        0.00
75%    429313.00     10486.00          13227.00         16762.00       16789.00      11.29    3922.00           1.00        0.00        0.00
max    517948.00     14574.00          17541.00         25153.00       25136.00     680.08 2234210.00           1.00        1.00        0.00

### feature selction for training

In [14]:
x = df_data.drop(['Trip ID', 'driver_action'], axis=1)

In [15]:
y = df_data['driver_action']

In [17]:
x.shape

(1555718, 8)

In [18]:
y.shape

(1555718,)

In [20]:
y.value_counts()

1    1529847
0      25871
Name: driver_action, dtype: int64

**1 is for rejected and 0 is for accepted**

**Train test split**

In [22]:
X_train, X_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=0)

In [ ]:
**
from xGBClassifierModel import XGBClassifierModel
boost_model = XGBClassifierModel(X_train, X_test, y_train, y_test)
clf3, boost_acc_arr, boost_loss_arr = boost_model.train(folds = 2)
fd3 = boost_model.get_feature_importance()
fig = plot_importance(fd3)

#### Using RandomForestClassifier

In [29]:
# create model variable
rnd = RandomForestClassifier()
  
# fit the model
fit_rnd = rnd.fit(X_train,y_train)
  
# checking the accuracy score
rnd_score = rnd.score(X_test,y_test) 
  
print('score of model is : ',rnd_score)
  
#pickle.dump(rnd, open("model.pkl", 'wb'))
#dump(rnd, 'model.joblib')

In [26]:
rnd_score

0.9832167742267246

### Using XGboost 

In [36]:
!pip install xgboost

  Using cached xgboost-1.6.2-py3-none-win_amd64.whl (125.4 MB)


In [40]:
# Define XGBoost Model
xgb = XGBClassifier(n_estimators=1000, learning_rate=0.05)
# We fit our model with our train data
xgb=xgb.fit(
    X_train, y_train,
    # That means if model don't improve it self in 5 rounds, it will stop learning
    # So you can save your time and don't overtrain your model.
    early_stopping_rounds=5,
    # We provide Test data's to evaluate model performance
    eval_set=[(X_test, y_test)],
    verbose=False
 )
# Then predict results from X_test data
pred_xgb = xgb.predict(X_test)

# See First 10 Predictions and They Actual Values
# print("Predicted:", pred_xgb[0:10])
print("Actual:", y_test[0:10])

Actual: 1004742    1
95574      1
1143161    1
744548     1
1199183    1
679978     1
989426     1
865656     1
1028013    1
1078509    1
Name: driver_action, dtype: int32


In [41]:
xg_score = xgb.score(X_test,y_test) 

In [42]:
xg_score

0.9835767361736045

**Both algorithm scores 98% of prediction on the given datasets**